### PR TITLE
Start central serialization methods

### LINKED ISSUE #1083: Warnings (?) caused by Serialization / Polymake
Running the test suite, I see this:
```
...
Test Summary: | Pass  Total
Serialization |   23     23
	(in cleanup) can't create a new copy of /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/moae.sop: No such file or directory
polymake:  WARNING: could not rename temporary file /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/poly36163Taaaa0011 to destination /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/moae.sop: No such file or directory
	(in cleanup) can't create a new copy of /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/moae.sop: No such file or directory
polymake:  WARNING: could not rename temporary file /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/poly36163Taaaa0012 to destination /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/moae.sop: No such file or directory
	(in cleanup) can't create a new copy of /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/lp.poly: No such file or directory
polymake:  WARNING: could not rename temporary file /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/poly36163Taaaa0013 to destination /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/lp.poly: No such file or directory
	(in cleanup) can't create a new copy of /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/lp.poly: No such file or directory
polymake:  WARNING: could not rename temporary file /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/poly36163Taaaa0014 to destination /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/lp.poly: No such file or directory
	(in cleanup) can't create a new copy of /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/nfsquare.fan: No such file or directory
polymake:  WARNING: could not rename temporary file /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/poly36163Taaaa0015 to destination /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/nfsquare.fan: No such file or directory
	(in cleanup) can't create a new copy of /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/nfsquare.fan: No such file or directory
polymake:  WARNING: could not rename temporary file /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/poly36163Taaaa0016 to destination /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/nfsquare.fan: No such file or directory
	(in cleanup) can't create a new copy of /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/square.poly: No such file or directory
polymake:  WARNING: could not rename temporary file /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/poly36163Taaaa0017 to destination /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/square.poly: No such file or directory
	(in cleanup) can't create a new copy of /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/square.poly: No such file or directory
polymake:  WARNING: could not rename temporary file /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/poly36163Taaaa0018 to destination /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/square.poly: No such file or directory
	(in cleanup) can't create a new copy of /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/posorth.cone: No such file or directory
polymake:  WARNING: could not rename temporary file /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/poly36163Taaaa0019 to destination /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/posorth.cone: No such file or directory
	(in cleanup) can't create a new copy of /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/posorth.cone: No such file or directory
polymake:  WARNING: could not rename temporary file /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/poly36163Taaaa0020 to destination /private/var/folders/9c/w7mlr03x2yd67g7n6384lvlc0000gp/T/jl_El4HRp/posorth.cone: No such file or directory
#I  Forcing finiteness test
Test Summary:     | Pass  Total
SimplicialComplex |   26     26
...
```

CC @lkastner @benlorenz

### PR DESCRIPTION
**Pull request by lkastner**
Since I started this PR it has been refactored after feedback from various people. Now the serialization is done by creating functions `save_intern` and `load_intern` for the different types we want to serialize and these functions pass on structs `SerializerState` and `DeserializerState` for recording the state, to allow backrefs and recording the context.

Obviously I cannot create these methods for everything in Oscar all by myself, so I hope that this is acceptable as a blueprint/draft.

### CODE REVIEW

File: src/PolyhedralGeometry/Serialization.jl [hunk 1]
```diff
@@ -1,170 +1,131 @@
...
-   if typename != "Cone<Rational>"
-      throw(ArgumentError("Loaded object is not of polymake type Cone, it has type " * typename))
-   end
-   return Cone(bigobject)
-end
 
 ##############################################################################
 """
-    save_polyhedron(Cone, String)
+    save(Cone, String)
```
Comment (fingolfin) [inline src/PolyhedralGeometry/Serialization.jl]: This does not match the code below, nor does it fit into the conventions for docstrings. Should be like this (also applies elsewhere)
```suggestion
    save(P::Polyhedron, filename::String)
```

File: src/PolyhedralGeometry/Serialization.jl [hunk 2]
```diff
@@ -1,170 +1,131 @@
...
    bigobject = pm_object(C)
    Polymake.save_bigobject(bigobject, filename)
 end
 
-"""
-    load_cone(String)
-
-Load a cone stored in JSON format, given the filename as input.
-"""
-function load_cone(filename::String)
```
Comment (fingolfin) [inline src/PolyhedralGeometry/Serialization.jl:18]: An alternative approach would be to replace `load_cone(str)` by `load(Cone, str)` (where `Cone` literally is the type, passed in as an argument). I.e.
```
function load(::Type{Cone}, filename::String)
```

File: src/PolyhedralGeometry/Serialization.jl (hunk 1 above)
Comment (lkastner) [reply src/PolyhedralGeometry/Serialization.jl]: Thanks

File: src/PolyhedralGeometry/Serialization.jl (hunk 2 above)
Comment (lkastner) [reply src/PolyhedralGeometry/Serialization.jl:18]: Well, but then the user needs to provide what they think this object is anyway. I can do this though, if this is preferred.

**Comment by fieker** [comment]
On Wed, Jan 19, 2022 at 03:09:41PM -0800, Lars Kastner wrote:
> The existing serialization methods are refactored into two central methods `save` and `load` that recognize the datatype from the JSON file.
> - Add JSON as dependency
> - Currently I only do the polyhedral objects, but I clearly mark the places where one would place a hook for loading other software.
> - The `save` side is easy, as one only needs to define a `save` function for every object.
> 
> I am aware of that the return type of `load` will be very ambiguous, I have raised this issue myself and this was the reason for the previous structure of the code. I have been told though, that a central `save` and `load` are desired.

It is a start and I am glad. However, I still would like to understand/
diskuss how this can be made to work in a "more complicated" setting,
i.e. where parent objects are involved. The current syntax, I think,
cannot be made to work for that. In order to have a stackable system
that allows more complicated objects a context/ dictionary/ ... has to
be included somewhere I think.

Consider a polynomial over some number field, start with a single
polynomial.
To serialize:
 - the number field needs to be involved
 - then I can serialize the degree
 - then a number of elements of the field
   A number field element is potentially a polynomial over a smaller
   field, thus, ...
   In particular, when reading the polynomial in, all coefficients need
   to be in the identical field, hence the field has to be part of the
   reading interface.

As a model I can see s.th. like
 - create a serialization context (like an ID-Dict)
 - for every object in this call
   see if it is already serialized, then send a reference (using the
   dict)
   else send the object and store the info in the dict.
For this to work, every serialization/ deserialization function needs to
accept the context/ dict as an argument... this is why we should discuss
this BEFORE writing many such functions...

File: src/PolyhedralGeometry/Serialization.jl (hunk 2 above)
Comment (fingolfin) [reply src/PolyhedralGeometry/Serialization.jl:18]: I can't say what is "preferred" as I still don't know what the big plan is for using these things. All I am saying is: if one is willing to write `load_cone(str)` then one should also be willing to accept `load(Cone, str)`, but that one has the advantage of being easier to use in generic code.

File: src/PolyhedralGeometry/Serialization.jl [hunk 3]
```diff
@@ -1,170 +1,211 @@
...
 Load a cone stored in JSON format, given the filename as input.
 """
-function load_cone(filename::String)
-   bigobject = Polymake.load_bigobject(filename)
-   typename = Polymake.type_name(bigobject)
-   if typename != "Cone<Rational>"
-      throw(ArgumentError("Loaded object is not of polymake type Cone, it has type " * typename))
-   end
-   return Cone(bigobject)
+function load(as::Type{Cone}, filename::String)
```
Comment (fingolfin) [inline src/PolyhedralGeometry/Serialization.jl]:
```suggestion
function load(::Type{Cone}, filename::String)
```

File: docs/src/PolyhedralGeometry/cones.md [hunk 4]
```diff
@@ -43,19 +43,14 @@ Objects of type `Cone` can be saved to a file and loaded from a file in the
 following way:
 ```@repl oscar
 C = positive_hull([1 0; 0 1])
-save_cone(C, "C.cone")
-CC = load_cone("C.cone")
+save(C, "C.cone")
+CC = load("C.cone")
```
Comment (fingolfin) [inline docs/src/PolyhedralGeometry/cones.md:47]: Should this perhaps be
```suggestion
CC = load(Cone, "C.cone")
```
?

**Comment by fieker** [comment]
Will there be a version with a stream or file passed in? To allow multiple
objects in the same file? This would be necessary for building and
serialising complex structures

File: docs/src/PolyhedralGeometry/cones.md (hunk 4 above)
Comment (lkastner) [reply docs/src/PolyhedralGeometry/cones.md:47]: So the question is, whether we can have both, i.e. `load(filename)` and `load(Type, filename)`. The latter for places where type inference matters for performance reasons, the former for convenience, i.e. inexperienced users, we don't have to explain to people how to figure out what is in the file or how to tell their colleagues. I now added both, and I think it doesn't hurt to have `load(filename)` for now. If it becomes too unmanageable we can always eliminate it. Is there some strong reason against that I don't see?

File: src/PolyhedralGeometry/Serialization.jl (hunk 3 above)
Comment (lkastner) [reply src/PolyhedralGeometry/Serialization.jl]: Thanks, makes sense, wasn't clear to me.

**Comment by lkastner** [comment]
> Will there be a version with a stream or file passed in? To allow multiple objects in the same file? This would be necessary for building and serialising complex structures

Yes, this is the long-term structure I have in mind. Currently it detects that the objects passed in come from polymake via the namespace parameter. So a new structure can either leave out this parameter or introduce its own namespace. We would probably have a oscar namespace and then use another parameter to detect the type, then it should be possible to recurse.

Nevertheless for now I just wanted to enable reading and writing for the polyhedral objects via Polymake.jl but I tried to keep everything as flexible as possible by modularizing with these future developments in mind. However there are still many things to discuss, as you mentioned.

**Comment by micjoswig** [comment]
> However, I still would like to understand/ diskuss how this can be made to work in a "more complicated" setting, i.e. where parent objects are involved.  ...

This was discussed (in some detail) at the recent polymake workshop.  The scenario we were considering is (still) "convex hulls over number fields"; so that should be general enough.  Here is the rough sketch.

(1) OSCAR types without polymake equivalent will live in a separate name space. Objects from several name spaces may live in the same file together, but more often there is one top level object from one name space.  In the more distant future it is conceivable to have further name spaces (beyond polymake and OSCAR).

(2) We will allow for pointers.  These may come in two flavors: relative via paths in the JSON tree (this includes, e.g., parents) and absolute (via URI).

Ewgenij created a sketch; will be discussed, then implemented.  The goal is to make the existing data base function accessible to all objects with such a serialization.

**Comment by lkastner** [comment]
@fingolfin @benlorenz Do you think this is ok for now? I tried to follow along the lines of @fingolfin 's writeup (Thanks a lot btw :wink: ). I don't want to make this PR any larger, so if you agree, I would merge and then work on further parts in separate PRs.

File: src/Serialization/Fields.jl [hunk 5]
```diff
@@ -0,0 +1,50 @@
+################################################################################
+# non-fmpz variant
+function save_intern(s::SerializerState, F::Nemo.GaloisField)
+    return Dict(
+        :characteristic => UInt64(characteristic(F))
+    )
+end
+
+function load_intern(s::DeserializerState, F::Type{Nemo.GaloisField}, dict::Dict)
+    return F(UInt64(dict[:characteristic]))
```
Comment (fingolfin) [inline src/Serialization/Fields.jl]: Note that `F === Nemo.GaloisField` here, so you could rewrite this code as saying
```suggestion
function load_intern(s::DeserializerState, ::Type{Nemo.GaloisField}, dict::Dict)
    return Nemo.GaloisField(UInt64(dict[:characteristic]))
```
Similar elsewhere.

**Review by fingolfin — APPROVED** [review]
Needs a rebase. Do we want/need to discuss this further, @fieker @lkastner ? Or merge it and discuss later? Either way is fine by me.

**Comment by fingolfin** [comment]
(Before merging, a conflict needs to be resolved)

**Comment by lkastner** [comment]
Benjamin suggested to prefix things that come from (or go to) polymake with a symbol `:polymake`. Another idea was to further automate serialization of our objects by just looping over their properties, but the latter part seems to fail (for now). I will implement the first suggestion and then merge.

**Review by benlorenz — APPROVED** [review]

**Pull request merged**